### ISSUE TITLE
Explore the idea of 'currency chemicals'

### ISSUE THREAD

**Issue by ukemi** [geneontology/go-ontology#16264]
On the RO call we brought up the idea of subdividing the types of chemicals in a reaction into a subgroup that represents chemicals that a relatively specific for a reaction and chemicals that are used over and over again, for example glucose versus NADH. These 'currency chemicals' would be different kinds of participants in reactions than other chemicals such as substrates and products. Using the prototype logical definitions to Rhea reactions, we should be able to derive a graph or table that indicates the frequency that a chemical is used in the universe set of reactions. For example H+ is used in 75% of all biochemical reactions while glucose is used in 0.4%. 

@deustp01 thoughts or ideas about this?

@goodb do you think this could be easily done?

**Comment by deustp01** [comment]
We have done something similar in our pathway display. When it is zoomed out, "housekeeping" molecules are grayed out leaving "specific" ones visible. We have identified the two kinds of molecules by building a housekeeping list by hand (NAD(P)(H), ATP, ADP, H2O - the usual suspects) and it does clean up the displays, with mostly user-friendly results. Processes whose purpose is ATP generation or ion transport look odd by default but can be manually tweaked to look OK.

All of that leaves me thinking that a list of chemicals ranked by frequency of usage would be useful to look at, and it might let us see ways of subdividing them, as well as the exceptions and traps. E.g., is there a threshold frequency below which a chemical can reliably be identified as substrate / product? Are there chemicals that can have both sorts of roles depending on the process and maybe depending on the taxon in which the reaction is occurring? My hunch is that these distinctions, much like precise pathway boundaries, will turn out to be fuzzy or arbitrary in at least some cases, but we may be able to see conditions where it would be safe and useful to apply them. Definitely worth a look.

**Comment by cmungall** [comment]
ccing @LEHunter who proposed the currency model. Link to broader ticket: https://github.com/oborel/obo-relations/issues/244

**Comment by LEHunter** [comment]
Might also make sense to pay attention to counting the times a reactant is the input to a directional reaction vs. output of one.  So, e.g. ATP is a very common input and ADP is a very common output, but rarely vice versa.  Which allows one to identify the situations where ATP actually *is* the substrate, i.e. during its synthesis.

Relating to the broader ticket, it means we don't have to find a definition for agent, enabler, main participant, etc.   The main participants are those that are not currency metabolites, e.g. are inputs, catalysts or outputs in a small proportion of reactions.

**Comment by amorgat** [comment]
@deustp01
We can already address the following query using [Rhea SPARQL endpoint](https://sparql.rhea-db.org/sparql): **_Select all reaction participants and count their occurrence in Rhea reactions_** 

PPREFIX rh:<http://rdf.rhea-db.org/>
SELECT ?reactionParticipantAc ?chebi
       (count(distinct ?reaction) as ?reactionCount) 
       ?reactionParticipantName 
WHERE {
  ?reaction rdfs:subClassOf rh:Reaction .
  ?reaction rh:status rh:Approved .
  ?reaction rh:side ?reactionSide .
  ?reactionSide rh:contains ?participant .  
  ?participant rh:compound ?compound .
  OPTIONAL { ?compound rh:chebi ?chebi . }
  ?compound rh:name ?reactionParticipantName .
  ?compound rh:accession ?reactionParticipantAc .
}
GROUP BY ?reactionParticipantAc ?chebi ?reactionParticipantName
ORDER BY DESC(?reactionCount)

**Comment by goodb** [comment]
Apologies for accidentally closing...  If its useful, I could give a count like @amorgat 's RHEA query that would indicate frequency of Left Hand Side versus Right Hand side in Reactome.  If we had an endpoint running with the GO-CAMs converted from Reactome we could do this in SPARQL right now (ping @lpalbou ).

I suspect the edge cases will be a little tricky here.  Ideally there would be some natural feature of the molecules that would indicate there nature as currency or housekeeping chemicals, but sparing that, I suspect a manually generated list like @deustp01 mentioned Reactome has is probably the way forward.  

(Attaching the RHEA results here as a csv for convenience)  
[rhea_reactant_counts.txt](https://github.com/geneontology/go-ontology/files/2297996/rhea_reactant_counts.txt)

**Comment by LEHunter** [comment]
Knowing the directionality of reactions isn't critical, the basic idea works without it.  Adding directionality allows one to additionally label currency metabolites as "main" products in reactions that produce them (rare) rather than consume them (common).

And what edge cases are you thinking of?  I bet the distribution of participates_in compounds would be very bimodal and easily separable.   The reason to do it this way is that there doesn't seem to be any natural feature of the molecule (or even of individual reactions) that discriminates the "main" participants.

**Comment by goodb** [comment]
Looking at the data, it does drop off very rapidly as you suggest and would expect.  There are only 34 compounds that show up in more than 100 rhea reactions and only 363 that show up in more than 10 (out of 9862 total compounds).  

The edge case I'm thinking of simply where to draw the line as without and natural feature this will be an arbitrary decision.  I'm not opposed as long as the 'currency' labeling is clearly useful directly in the context of the ontology.  How might it influence reasoning?  Or is this something that consumers of the data (e.g. people that want to make a display like Reactome's with very common molecules greyed out) should just compute for themselves and set their own arbitrary boundaries for?

**Comment by deustp01** [comment]
Bijay Jassal and Steve Jupe at Reactome identified a list of about 30 chemicals that often play accessory (housekeeping, trivial, currency) roles in Reactions and are thus unlikely to be good entities to use in GO as parts of pathway definitions or as pathway endpoints. Bijay has also generated the number of times each of these chemicals is annotated as a participant in a Reactome reaction. Here, this list is aligned with the list of chemicals used as participants in Rhea reactions generated last summer.
The ChEBI IDs for individual chemicals often differ between Rhea and Reactome; the Rhea usage is probably the correct one and developing an efficient process to find and fix these discrepancies is one of the goals of this whole project.
A few impressions. The 13 chemicals most extensively used in Rhea are all on the Reactome list – no surprises here. Sixteen of the other 17 Reactome chemicals are also listed by Rhea with usages ranging from four to 369. A surprising result is that Rhea has no entry for OH- (ChEBI:16234) that I could find. Looking at the chemicals widely used in Rhea or Reactome but not on the Reactome “housekeeping” list, I didn’t see any that looked like good candidates for housekeeping status – items like individual amino acids and fatty acids, for example. Still, the current Reactome list looks like a plausible starting point for a list that Reactome, GO, and Rhea might all use.
It also looks like there is some scope here for guidelines to distinguish reactions where a chemical is likely to have an accessory role from ones where its role is likely to be the focus of the reaction. NAD+/NADH, NADP+/NADPH, FAD/FADH2, FMN/FMNH2 when paired as inputs and outputs (either order) are likely to be accessories; when one is present without the other it is more likely to be a focus molecule. NTP input paired with NDP output likely indicates accessory roles; NDP input and NTP output suggest a focus role.
Here is the table - apologies for format.

	reactionParticipantAc	reactionCount	RHEA name			ChEBI ID	count	Reactome name
RHEA	CHEBI:15378	6544	H(+)		Reactome	15378	596	H+
RHEA	CHEBI:15377	4235	H2O		Reactome	15377	1175	H2O
RHEA	CHEBI:15379	1785	O2		Reactome	15379	365	O2
RHEA	CHEBI:57287	1051	CoA		Reactome	15346	192	CoA
RHEA	CHEBI:58349	986	NADP(+)		Reactome	18009	281	NADP+
RHEA	CHEBI:57783	981	NADPH		Reactome	13392	311	NADPH
RHEA	CHEBI:30616	876	ATP		Reactome	15422	1921	ATP
RHEA	CHEBI:57540	858	NAD(+)		Reactome	15846	143	NAD+
RHEA	CHEBI:57945	839	NADH		Reactome	16908	117	NADH
RHEA	CHEBI:33019	743	diphosphate		Reactome	29888	213	Ppi
RHEA	CHEBI:16526	715	CO2		Reactome	16526	142	CO2
RHEA	CHEBI:43474	702	phosphate		Reactome	18367	443	Pi
RHEA	CHEBI:456216	609	ADP		Reactome	16761	1462	ADP
RHEA	CHEBI:58223	369	UDP		Reactome	17659	91	UDP
RHEA	CHEBI:456215	301	AMP		Reactome	16027	178	AMP
RHEA	CHEBI:58189	98	GDP		Reactome	17552	682	GDP
RHEA	CHEBI:57692	70	FAD		Reactome	16238	120	FAD
RHEA	CHEBI:58307	66	FADH2		Reactome	17877	28	FADH2
RHEA	CHEBI:37565	54	GTP		Reactome	15996	1522	GTP
RHEA	CHEBI:57865	33	UMP		Reactome	16695	30	UMP
RHEA	CHEBI:58115	21	GMP		Reactome	17345	24	GMP
RHEA	CHEBI:46398	18	UTP		Reactome	15713	105	UTP
RHEA	CHEBI:61429	6	dGTP		Reactome	16497	10	dGTP
RHEA	CHEBI:57667	5	dADP		Reactome	16174	10	dADP
RHEA	CHEBI:61404	5	dATP		Reactome	16284	12	dATP
RHEA	CHEBI:58595	5	dGDP		Reactome	28862	6	dGDP
RHEA	CHEBI:57673	4	dGMP		Reactome	16192	9	dGMP
RHEA	CHEBI:58245	4	dAMP		Reactome	17713	13	dAMP
RHEA	CHEBI:61555	4	dUTP		Reactome	17625	5	dUTP
								
			no entry in Rhea?		Reactome	16234		OH-

**Comment by fabregat** [comment]
Same exercise for Reactome just in case it helps.

The attached file contains the number of times each ChEBI identifier is used in different reactions (for "Homo sapiens").

[Reactome_chemical_reactions_count.txt](https://github.com/geneontology/go-ontology/files/2662979/Reactome_chemical_reactions_count.txt)

**Comment by ukemi** [comment]
https://github.com/geneontology/go-ontology/pull/16797

**Comment by ukemi** [comment]
Rather than specify currency chemicals, we went the route of specifying the molecules that were important in metabolic processes by defining them as primary inputs and outputs. Molecules that are not primary inputs and outputs will therefore be transient intermediates and currency chemicals.

**Comment by LEHunter** [comment]
Sorry to hear this got closed without further discussion.   How did you solve the problem of systematically determining which inputs and outputs were primary?  Where is the definition of "primary" or "primary input" and "primary output"?

**Comment by ukemi** [comment]
Sorry Larry, I can reopen it if you want. The definitions are in RO. For metabolic reactions in GO, the distinction was clear because most of them are generic and the key players are named in the process. So glucose catabolic process is a catabolic process that has a primary input of glucose.

**Comment by LEHunter** [comment]
URLs for RO definitions?   Not seeing them…  Surprising it’s not controversial after all the sturm und drang about it in the past.

**Comment by cmungall** [comment]
I'm not sure why they haven't propagated to OLS, but you can see them here:

http://purl.obolibrary.org/obo/RO_0004009
http://purl.obolibrary.org/obo/RO_0004008

To reinforce David's point, it sounds a bit circular, the GO process terms are based on the objectives of the process, so the relations reflect this. So typically ATP ADP etc don't reflect the "goals", except when they do, so there is no rigid classification based on the chemical entity. But this is an ongoing discussion over on the GO list, the currency non-currency distinction is definitely useful for thinking about how we structure the ontology.

**Comment by LEHunter** [comment]
Ugh.  Processes don’t have goals.  These definitions are uninterpretable.  What is the “goal” of TCA?  ATP? NADH?  FADH2? α-Ketoglutarate? Glutamine? Serine?  

*big sigh*

**Comment by addiehl** [comment]
The "goal' is the interesting outcome of the process from a human perspective.  Mostly joking, but not entirely.

**Comment by addiehl** [comment]
I think if GO BP are realizations of BFO functions, those functions are selected by evolution with particular goals.  Per Spear, Ceusters, and Smith "the function of a thing is, approximately, what it is supposed to do. This typically involves both what is popularly called a goal or end of some sort and a way of achieving that end. A thing is functioning well when it does what it is supposed to do reliably or effectively and in the, for it, appropriate way. A thing is malfunctioning when it does what it is supposed to do poorly or not at all. The function of the heart is to pump blood. When this function is realized – when the heart is pumping – then blood flows through the circulatory system. This is a consequence of the realization of the heart’s function. A healthy or well-functioning heart does this well. An unhealthy or malfunctioning heart does this poorly...Talk of functions is thus in some sense teleological (involves reference to goals or ends or purposes) and normative (reference to functions serves as the basis of certain sorts of judgments of good and bad, better and worse)." http://ontology.buffalo.edu/smith/articles/Functions-in-BFO.pdf

**Comment by addiehl** [comment]
One can imagine the creation of a biological function ontology to match up with GO BP, although it's not a likely possibility, so we have to make do with the process classes we have to work with.

**Comment by cmungall** [comment]
> One can imagine the creation of a biological function ontology to match up with GO BP, although it's not a likely possibility, so we have to make do with the process classes we have to work with.

Not likely and not required, this would be a completely superfluous shadow ontology.

However, it would be useful to include an axiom in GO-plus stating that every GO BP is the realization multiple BFO functions

**Comment by cmungall** [comment]
> Ugh.  Processes don’t have goals.

Well, this is pretty much baked into the GO. The definition of the root BP node:

def: "A biological process represents a specific objective that the organism is genetically programmed to achieve. Biological processes are often described by their outcome or ending state, e.g., the biological process of cell division results in the creation of two daughter cells (a divided cell) from a single parent cell. A biological process is accomplished by a particular set of molecular functions carried out by specific gene products (or macromolecular complexes), often in a highly regulated manner and in a particular temporal sequence." [GOC:pdt]

See also Paul's chapter in the GO handbook.

We'd probably be better taking discussion of this to a new ticket since it is fundamental to GO and bigger than the original source ticket

But it is baked in to the GO and the existing BP biochemical branch, we should do a better job of communicating some of this and the ticket history...

> These definitions are uninterpretable.  What is the “goal” of TCA?  ATP? NADH?  FADH2? α-Ketoglutarate? Glutamine? Serine?

Yep, these don't have goals, but the process has evolved to do something. The role of these chemicals in the process may vary depending on the process.

Note that I think currency role is v useful, and chebi should include as has_role, this could also be useful for querying go

**Comment by addiehl** [comment]
Unfortunately the definition of biological process is simply terrible from an ontological perspective, since a process is not an objective nor represents an objective either in BFO and in the common sense of the words.  I did point out this revised definition to Barry a while back, for the shock value, though I guess he's got other battles to wage.  Per OBI and IAO, a planned process _achieves_ some objective specification; might be nice to come up with some biological equivalent.

Basically, this is the kind of issue that should be discussed at the OBO-core workshop.
